pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,427 @@
+// - request ack from candidates (update LeaseCandidate PingTime)
+// - finds the most optimal candidate and elect (update the Lease object)
+// Instead of keeping a map and lock on election, the state is
+// calculated every time by looking at the lease, and set of available candidates.
+// PingTime + electionDuration > time.Now: We just asked all candidates to ack and are still waiting for response
+// PingTime + electionDuration < time.Now: Candidate has not responded within the appropriate PingTime. Continue the election.
+// RenewTime + 5 seconds > time.Now: All candidates acked in the last 5 seconds, continue the election.
+func (c *Controller) reconcileElectionStep(ctx context.Context, leaseNN types.NamespacedName) (requeue time.Duration, err error) {
+	now := time.Now()
+
+	candidates, err := c.listAdmissableCandidates(leaseNN)
+	if err != nil {
+		return defaultRequeueInterval, err
+	} else if len(candidates) == 0 {
+		return noRequeue, nil
+	}
+	klog.V(6).Infof("Reconciling election for %s, candidates: %d", leaseNN, len(candidates))
+
+	// Check if an election is really needed by looking at the current lease and candidates
+	needElection, err := c.electionNeeded(candidates, leaseNN)
+	if !needElection {
+		return noRequeue, err
+	}
+	if err != nil {
+		return defaultRequeueInterval, err
+	}
+
+	canVoteYet := true
+	for _, candidate := range candidates {
+		if candidate.Spec.PingTime != nil && candidate.Spec.PingTime.Add(electionDuration).After(now) {
+			// continue waiting for the election to timeout
+			canVoteYet = false
+			continue
+		}
+		if candidate.Spec.RenewTime.Add(electionDuration).After(now) {
+			continue
+		}
+
+		if candidate.Spec.PingTime == nil ||
+			// If PingTime is outdated, send another PingTime only if it already acked the first one.
+			(candidate.Spec.PingTime.Add(electionDuration).Before(now) && candidate.Spec.PingTime.Before(candidate.Spec.RenewTime)) {
+			clone := candidate.DeepCopy()
+			clone.Spec.PingTime = &metav1.MicroTime{Time: time.Now()}
+			_, err := c.leaseCandidateClient.LeaseCandidates(clone.Namespace).Update(ctx, clone, metav1.UpdateOptions{})
+			if err != nil {
+				return noRequeue, err
+			}
+			if candidate.Spec.PingTime == nil {
+				canVoteYet = false
+			}
+		}
+	}
+	if !canVoteYet {
+		return defaultRequeueInterval, nil
+	}
+
+	// election is ongoing as long as unexpired PingTimes exist
+	for _, candidate := range candidates {
+		if candidate.Spec.PingTime == nil {
+			continue // shouldn't be the case after the above
+		}
+
+		if candidate.Spec.RenewTime != nil && candidate.Spec.PingTime.Before(candidate.Spec.RenewTime) {
+			continue // this has renewed already
+		}
+
+		// If a candidate has a PingTime within the election duration, they have not acked
+		// and we should wait until we receive their response
+		if candidate.Spec.PingTime.Add(electionDuration).After(now) {
+			// continue waiting for the election to timeout
+			return noRequeue, nil
+		}
+	}
```

Comment:
am not completely sure how this loop relates to the previous one. Looks like we recheck what we do above already.

Reply:
Follow-up.